File: src/scandoc/tui/events.py
"""
Application Layer Event Bus for scanDOC TUI and multi-interface event distribution.
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Callable, Dict, List, Optional


class EventType(str, Enum):
    DOCUMENT_IMPORTED = "document.imported"
    PROCESSING_STARTED = "processing.started"
    STAGE_STARTED = "stage.started"
    STAGE_COMPLETED = "stage.completed"
    PROCESSING_COMPLETED = "processing.completed"
    PROCESSING_FAILED = "processing.failed"
    JOB_STATUS_CHANGED = "job.status_changed"
    MODEL_STATUS_CHANGED = "model.status_changed"
    LOG_EMITTED = "log.emitted"
    SERVER_STATUS_CHANGED = "server.status_changed"


@dataclass
class AppEvent:
    event_type: EventType
    timestamp: datetime = field(default_factory=datetime.now)
    payload: Dict[str, Any] = field(default_factory=dict)
    message: Optional[str] = None


class EventBus:
    """
    Decoupled Event Bus enabling core engine events to update application state,
    job manager, logging infrastructure, and presentation layers without circular dependencies.
    """

    def __init__(self):
        self._subscribers: Dict[EventType, List[Callable[[AppEvent], None]]] = {}

    def subscribe(self, event_type: EventType, callback: Callable[[AppEvent], None]) -> None:
        if event_type not in self._subscribers:
            self._subscribers[event_type] = []
        self._subscribers[event_type].append(callback)

    def publish(self, event: AppEvent) -> None:
        callbacks = self._subscribers.get(event.event_type, [])
        for callback in callbacks:
            try:
                callback(event)
            except Exception as e:
                pass


default_event_bus = EventBus()

File: internal/tui/events/events.go
package events

import (
	"sync"
	"time"

	tea "github.com/charmbracelet/bubbletea"
)

type EventType string

const (
	EventDocumentImported    EventType = "document.imported"
	EventProcessingStarted   EventType = "processing.started"
	EventStageStarted        EventType = "stage.started"
	EventStageCompleted      EventType = "stage.completed"
	EventProcessingCompleted EventType = "processing.completed"
	EventProcessingFailed    EventType = "processing.failed"
	EventJobStatusChanged    EventType = "job.status_changed"
	EventModelStatusChanged  EventType = "model.status_changed"
	EventLogEmitted          EventType = "log.emitted"
	EventServerStatusChanged EventType = "server.status_changed"
)

type AppEvent struct {
	Type      EventType      `json:"type"`
	Timestamp time.Time      `json:"timestamp"`
	Payload   map[string]any `json:"payload"`
	Message   string         `json:"message"`
}

// Bubble Tea Message Wrapper
type AppEventMsg AppEvent

type EventBus struct {
	mu          sync.RWMutex
	subscribers map[EventType][]func(AppEvent)
}

func NewEventBus() *EventBus {
	return &EventBus{
		subscribers: make(map[EventType][]func(AppEvent)),
	}
}

func (b *EventBus) Subscribe(eventType EventType, callback func(AppEvent)) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.subscribers[eventType] = append(b.subscribers[eventType], callback)
}

func (b *EventBus) Publish(event AppEvent) {
	b.mu.RLock()
	subs := b.subscribers[event.Type]
	callbacks := make([]func(AppEvent), len(subs))
	copy(callbacks, subs)
	b.mu.RUnlock()

	for _, cb := range callbacks {
		func() {
			defer func() { recover() }()
			cb(event)
		}()
	}
}

// EmitCmd returns a tea.Cmd for sending an event into Bubble Tea model Update
func EmitCmd(event AppEvent) tea.Cmd {
	return func() tea.Msg {
		if event.Timestamp.IsZero() {
			event.Timestamp = time.Now()
		}
		return AppEventMsg(event)
	}
}

var DefaultEventBus = NewEventBus()
